### PR TITLE
Channel wise distillation

### PR DESCRIPTION
**Pull request by lkdci**
Implementation of "Channel-wise Knowledge Distillation for Dense Prediction", [paper](https://arxiv.org/abs/2011.13256)

This PR includes:
* Wrapper loss `SegKDLoss` for a combo loss of an objective classification loss and a KD loss.
* Channel wise distillation implementation `CWDKlDivLoss`
* General recipe for KD training on Cityscapes dataset
* Up to date SOTA results for the DDRNet models family using the CW distillation, Updated the model zoo, pretrained models uri and pretrained models test.
*  Refactored DDRNet models to inherit `SegmentationModule` to follow the unified format of segmentation models in this repo. Flag for using aux heads is now `use_aux_heads` instead of `aux_head`, inheriting the base segmentation module also standardize the removal of aux heads before deployment, which wasn't applied before.

CWD mIoU results compared to paper and SG results, for the DDRNet models:

| Model                         |  paper mIoU | SuperGradients mIoU  | CWD mIoU  |
| -------------------- |  ------------ | ---------------------- | -----------  |
| DDRNet23-Slim        |  77.8              | 78.01                              | **79.41**     |
| DDRNet23                 |  79.5              | 80.26                             | **81.48**    |

### CODE REVIEW

File: src/super_gradients/training/models/segmentation_models/ddrnet.py [hunk 1]
```diff
@@ -369,15 +372,15 @@ def __init__(
                 strides=spp_strides,
             )
 
-            if self.aux_head:
-                self.seghead_extra = SegmentHead(highres_planes, head_width, num_classes, 8, inter_mode=self.segmentation_inter_mode)
-
             self.final_layer = SegmentHead(highres_planes * layer5_bottleneck_expansion, head_width, num_classes, 8, inter_mode=self.segmentation_inter_mode)
 
+            if self.use_aux_heads:
```
Comment (lkdci) [inline src/super_gradients/training/models/segmentation_models/ddrnet.py:377]: Changing the order allows loading pretrained checkpoints with auxiliary heads, to a model without auxiliary heads

File: src/super_gradients/training/Computer_Vision_Models_Pretrained_Checkpoints.md [hunk 2]
```diff
@@ -68,12 +68,17 @@ All the available models are listed in the column `Model name`.
...
 | STDC 1-Seg50          | stdc1_seg50       | Cityscapes | 512x1024   | 75.11 | **3.34ms**              | **30.12ms**                          |                    **27.54ms**                    |
 | STDC 1-Seg75          | stdc1_seg75       | Cityscapes | 768x1536   | 77.8  | **5.53ms**              | **32.490ms**                         |                     **43.88**                     |
 | STDC 2-Seg50          | stdc2_seg50       | Cityscapes | 512x1024   | 76.44 | **4.12ms**              | **30.94ms**                          |                    **32.03ms**                    |
 | STDC 2-Seg75          | stdc2_seg75       | Cityscapes | 768x1536   | 78.93 | **6.95ms**              | **33.89ms**                          |                   **54.48ms**                     |
 | RegSeg (exp48)        | regseg48          | Cityscapes | 1024x2048  | 78.15 | **12.03ms**             | **38.91ms**                          |                    **78.20ms**                    |
+| DDRNet 23 slim - CWD  | ddrnet_23_slim    | Cityscapes | 1024x2048  | 79.41 | **5.74ms**              | **32.01ms**                          |                    **45.18ms**                    |
```
Comment (shaydeci) [inline src/super_gradients/training/Computer_Vision_Models_Pretrained_Checkpoints.md:77]: We can just keep the best checkpoints and simply state in the recipe that the ious were reached with the KD ones.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Single comment about the table, awesome job!

File: src/super_gradients/training/Computer_Vision_Models_Pretrained_Checkpoints.md (hunk 2 above)
Comment (lkdci) [reply src/super_gradients/training/Computer_Vision_Models_Pretrained_Checkpoints.md]: Done [update model-zoo](https://github.com/Deci-AI/super-gradients/pull/731/commits/87192ff439dd9afbde5bea1607025f7360f7d6e0)

**Comment by lkdci** [comment]
Thanks for the review @shaydeci , Addressed

**Comment by lkdci** [comment]
@shaydeci , As discussed offline changed the note about ddrnet models training in model zoo
[update note](https://github.com/Deci-AI/super-gradients/pull/731/commits/6f74257a70b4aa3b25f83cf6fac0a039f3fdeca4)

**Review by shaydeci — APPROVED** [review]
LGTM

File: src/super_gradients/training/losses/seg_kd_loss.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+from super_gradients.training.models.kd_modules.kd_module import KDOutput
+
+
+class SegKDLoss(_Loss):
+    """
+    Wrapper loss for semantic segmentation KD.
+    This loss includes two loss components, `ce_loss` i.e CrossEntropyLoss, and `kd_loss` i.e `CWDKlDivLoss`.
+    """
+
+    def __init__(self, kd_loss: _Loss, ce_loss: _Loss, weights: Union[tuple, list], kd_loss_weights: Union[tuple, list]):
```
Comment (BloodAxe) [inline src/super_gradients/training/losses/seg_kd_loss.py:15]: SegKDLoss uses _Loss as type hints. It is not recommended to use private types (No guarantee _Loss will persist in upcoming releases) and perhaps nn.Module would be a better fit here since you don't use anything specific from _Loss class AFAIK

File: src/super_gradients/training/losses/cwd_loss.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+from typing import Optional
+
+from torch.nn.modules.loss import _Loss
+import torch.nn as nn
+import torch
+
+
+class CWDKlDivLoss(_Loss):
```
Comment (BloodAxe) [inline src/super_gradients/training/losses/cwd_loss.py:8]: CWDKlDivLoss is a bit hard to read and I wonder why not ChannelWiseKnowledgeDistillationLoss ?

File: src/super_gradients/training/losses/seg_kd_loss.py (hunk 3 above)
Comment (lkdci) [reply src/super_gradients/training/losses/seg_kd_loss.py]: Replaced to nn.Module [CR fixes](https://github.com/Deci-AI/super-gradients/pull/731/commits/59abdf6d228f14d71a74411ad85dac745bdd2efb)

File: src/super_gradients/training/losses/cwd_loss.py (hunk 4 above)
Comment (lkdci) [reply src/super_gradients/training/losses/cwd_loss.py]: Renamed [CR fixes](https://github.com/Deci-AI/super-gradients/pull/731/commits/59abdf6d228f14d71a74411ad85dac745bdd2efb)

**Comment by BloodAxe** [comment]
LGTM

**Review by BloodAxe — APPROVED** [review]
